From 3ae75a56067c8d2e14521622a66dcb0f7baeb4a9 Mon Sep 17 00:00:00 2001
From: Akhil Narang <akhilnarang.1999@gmail.com>
Date: Tue, 26 Jun 2018 16:55:14 +0530
Subject: [PATCH 292/296] Fix: ConnectivityManager is initialized with a wrong
 context

Symptom:
BT tethering was failed due to security exception. The error message
said it's requested by "android" package without system uid.

Root cause:
ActivityThread#setHttpProxy uses a system context to get
a ConnectivityManager instance. Unexpectedly, ConnectivityManager with
the system context is cached to the system service registry when it's
not cached yet. The wrong context is always reused when getting
ConnectivityManager.

Solution:
Use an Application context to get ConnectivityManager if it can be used.
---
 core/java/android/app/ActivityThread.java | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/core/java/android/app/ActivityThread.java b/core/java/android/app/ActivityThread.java
index 57c9386dd2d..dce037a16e8 100644
--- a/core/java/android/app/ActivityThread.java
+++ b/core/java/android/app/ActivityThread.java
@@ -987,7 +987,8 @@ public final class ActivityThread {
         }
 
         public void setHttpProxy(String host, String port, String exclList, Uri pacFileUrl) {
-            final ConnectivityManager cm = ConnectivityManager.from(getSystemContext());
+            final ConnectivityManager cm = ConnectivityManager.from(
+                getApplication() != null ? getApplication() : getSystemContext());
             final Network network = cm.getBoundNetworkForProcess();
             if (network != null) {
                 Proxy.setHttpProxySystemProperty(cm.getDefaultProxy());
-- 
2.11.0

